File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/mapper/BookUsingMapper.java
package ru.elenacherev.librarymanager.mapper;

import ru.elenacherev.librarymanager.api.dto.BookUsing;
import ru.elenacherev.librarymanager.domain.entity.BookEntity;
import ru.elenacherev.librarymanager.domain.entity.BookUsingEntity;
import ru.elenacherev.librarymanager.domain.entity.ReaderEntity;

public class BookUsingMapper {
    public static BookUsing map(BookUsingEntity entity) {
        return BookUsing.builder()
                .bookUsingId(entity.getBookUsingId())
                .startDate(entity.getStartDate())
                .endDate(entity.getEndDate())
                .bookId(entity.getBook()!=null
                			?entity.getBook().getBookId()
                			:0
                		)
                .readerId(entity.getReader().getReaderId())
                .build();
    }

    public static BookUsingEntity map(
    		BookUsing dto,
    		ReaderEntity reader,
    		BookEntity book
    ) {
        return map(new BookUsingEntity(), dto, reader, book);
    }

    public static BookUsingEntity map(
    		BookUsingEntity entity,
    		BookUsing dto, 
    		ReaderEntity reader,
    		BookEntity book
    ) {
        entity.setStartDate(dto.getStartDate());
        entity.setEndDate(dto.getEndDate());
        entity.setBook(book);
        entity.setReader(reader);
        return entity;
    }
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/mapper/BookUsingMapper.kt
package ru.elenacherev.librarymanager.mapper

import ru.elenacherev.librarymanager.api.dto.BookUsing
import ru.elenacherev.librarymanager.domain.entity.BookEntity
import ru.elenacherev.librarymanager.domain.entity.BookUsingEntity
import ru.elenacherev.librarymanager.domain.entity.ReaderEntity

fun BookUsingEntity.map() = BookUsing(
    bookUsingId = bookUsingId,
    startDate = startDate,
    endDate = endDate,
    bookId = book.bookId!!,
    readerId = reader.readerId!!
)

fun BookUsing.map(
    book: BookEntity,
    reader: ReaderEntity
) = BookUsingEntity(
    startDate = startDate,
    endDate = endDate,
    book = book,
    reader = reader
)

fun BookUsing.map(
    entity: BookUsingEntity,
    book: BookEntity,
    reader: ReaderEntity
) = entity.also {
    it.startDate = startDate
    it.endDate = endDate
    it.book = book
    it.reader = reader
}
